evm/src/generation/state.rs

```diff
@@ -0,0 +1,118 @@
+use ethereum_types::U256;
+use plonky2::field::types::Field;
+
+use crate::cpu::columns::NUM_CPU_COLUMNS;
+use crate::cpu::kernel::aggregator::combined_kernel;
+use crate::cpu::kernel::assembler::Kernel;
+use crate::generation::memory::MemoryState;
+use crate::logic::{Op, Operation};
+use crate::memory::memory_stark::MemoryOp;
+use crate::{keccak, logic};
+
+#[derive(Debug)]
+pub(crate) struct GenerationState<F: Field> {
+    pub(crate) kernel: Kernel,
+
+    pub(crate) cpu_rows: Vec<[F; NUM_CPU_COLUMNS]>,
+    pub(crate) current_cpu_row: [F; NUM_CPU_COLUMNS],
+
+    pub(crate) current_context: usize,
+    pub(crate) memory: MemoryState<F>,
+
+    pub(crate) keccak_inputs: Vec<[u64; keccak::keccak_stark::NUM_INPUTS]>,
+    pub(crate) logic_ops: Vec<logic::Operation>,
+}
+
+impl<F: Field> GenerationState<F> {
+    /// Compute logical AND, and record the operation to be added in the logic table later.
+    #[allow(unused)] // TODO: Should be used soon.
+    pub(crate) fn and(&mut self, input0: U256, input1: U256) -> U256 {
+        self.logic_op(Op::And, input0, input1)
+    }
+
+    /// Compute logical OR, and record the operation to be added in the logic table later.
+    #[allow(unused)] // TODO: Should be used soon.
+    pub(crate) fn or(&mut self, input0: U256, input1: U256) -> U256 {
+        self.logic_op(Op::Or, input0, input1)
+    }
+
+    /// Compute logical XOR, and record the operation to be added in the logic table later.
+    #[allow(unused)] // TODO: Should be used soon.
+    pub(crate) fn xor(&mut self, input0: U256, input1: U256) -> U256 {
+        self.logic_op(Op::Xor, input0, input1)
+    }
+
+    /// Compute logical AND, and record the operation to be added in the logic table later.
+    pub(crate) fn logic_op(&mut self, op: Op, input0: U256, input1: U256) -> U256 {
+        let operation = Operation::new(op, input0, input1);
+        let result = operation.result;
+        self.logic_ops.push(operation);
+        result
```

Comment:
Logic ops (also arithmetic) rows could potentially be reused, if an operation is executed multiple times with identical inputs. Don’t worry about this now, but potentially a thing to do in a future change.